### ISSUE TITLE
Bug:     Cannot find module '@jsprismarine/bedrock-data' from 'node_modules/@jsprismarine/prismarine/dist/block/Block.cjs.cjs'

### ISSUE THREAD

**Issue by vincss** [JSPrismarine/JSPrismarine#1683]
**Describe the bug**
I'm trying to update mcsleepingserverstarter with the latest version of jsprismarine, with that version I'm having an issue with the bedrock-data : 
Error [ERR_REQUIRE_ESM]: require() of ES Module E:\workspace\_externals\mcsleepingserverstarter\node_modules\@jsprismarine\bedrock-data\dist\index.js from E:\workspace\_externals\mcsleepingserverstarter\node_modules\@jsprismarine\prismarine\dist\block\Block.cjs.cjs not supported.

**To Reproduce**
Git clone https://github.com/vincss/mcsleepingserverstarter/tree/feature/bedrock_JSPrismarine_update?tab=readme-ov-file
Be sure to be on branch feature/bedrock_JSPrismarine_update
npm install
npm start

**Expected behavior**
The bedrock server starts.

**Screenshots**

**Desktop (please complete the following information):**
 - OS: Windows 11
 - Node 20

**Additional context**

**Comment by filiphsps** [comment]
Looks like we need to move bedrock-data over to the newer architecture @enricoangelon

**Comment by filiphsps** [comment]
Keep this open till a release is pushed and we've verified that it's fixed.

(frankly we should have a few automated tests for cjs, though that would also imply we support it which we kind of don't).

**Comment by filiphsps** [comment]
Alright @vincss try with version `0.8.0` and see if that works! (Should be available in a few minutes)

**Comment by vincss** [comment]
Hi,
Thanks for your quick anwser. 

I think it's getting a step further, I've got another error : 
`
E:\workspace\_externals\mcsleepingserverstarter\node_modules\@jsprismarine\prismarine\dist\network\handler\InventoryTransactionHandler.cjs.cjs:5
const minecraft = require('@jsprismarine/minecraft');
                  ^

Error [ERR_REQUIRE_ESM]: require() of ES Module E:\workspace\_externals\mcsleepingserverstarter\node_modules\@jsprismarine\minecraft\dist\index.cjs.js from E:\workspace\_externals\mcsleepingserverstarter\node_modules\@jsprismarine\prismarine\dist\network\handler\InventoryTransactionHandler.cjs.cjs not supported.
index.cjs.js is treated as an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which declares all .js files in that package scope as ES modules.
Instead either rename index.cjs.js to end in .cjs, change the requiring code to use dynamic import() which is available in all CommonJS modules, or change "type": "module" to "type": "commonjs" in E:\workspace\_externals\mcsleepingserverstarter\node_modules\@jsprismarine\minecraft\package.json to treat all .js files as CommonJS (using .mjs for all ES modules instead).

    at Module.<anonymous> (E:\workspace\_externals\mcsleepingserverstarter\node_modules\@jsprismarine\prismarine\dist\network\handler\InventoryTransactionHandler.cjs.cjs:5:19) {
  code: 'ERR_REQUIRE_ESM'
}
`

Also, maybe it would be good to add @jsprismarine/bedrock-data as dependencies in the @jsprismarine/prismarine .

**Comment by filiphsps** [comment]
> Hi, Thanks for your quick anwser.
> 
> I think it's getting a step further, I've got another error : ` E:\workspace_externals\mcsleepingserverstarter\node_modules@jsprismarine\prismarine\dist\network\handler\InventoryTransactionHandler.cjs.cjs:5 const minecraft = require('@jsprismarine/minecraft'); ^
> 
> Error [ERR_REQUIRE_ESM]: require() of ES Module E:\workspace_externals\mcsleepingserverstarter\node_modules@jsprismarine\minecraft\dist\index.cjs.js from E:\workspace_externals\mcsleepingserverstarter\node_modules@jsprismarine\prismarine\dist\network\handler\InventoryTransactionHandler.cjs.cjs not supported. index.cjs.js is treated as an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which declares all .js files in that package scope as ES modules. Instead either rename index.cjs.js to end in .cjs, change the requiring code to use dynamic import() which is available in all CommonJS modules, or change "type": "module" to "type": "commonjs" in E:\workspace_externals\mcsleepingserverstarter\node_modules@jsprismarine\minecraft\package.json to treat all .js files as CommonJS (using .mjs for all ES modules instead).
> 
> ```
> at Module.<anonymous> (E:\workspace\_externals\mcsleepingserverstarter\node_modules\@jsprismarine\prismarine\dist\network\handler\InventoryTransactionHandler.cjs.cjs:5:19) {
> ```
> 
> code: 'ERR_REQUIRE_ESM' } `
> 
> Also, maybe it would be good to add @jsprismarine/bedrock-data as dependencies in the @jsprismarine/prismarine .

Looks like `@jsprismarine/minecraft` & `@jsprismarine/protocol` didn't use `.cjs` for commonjs modules. That has been resolved in 996347038b55d4d08b08b1efdc451e89d7265b30 and a new release will be pushed out shortly.

**Comment by vincss** [comment]
Hi,
Again thanks for your quick answer, now it fails a bit further again : 
`E:\workspace\_externals\mcsleepingserverstarter\node_modules\@jsprismarine\prismarine\dist\ban\BanManager.cjs.cjs:6
const minifyJson = require('strip-json-comments');
                   ^

Error [ERR_REQUIRE_ESM]: require() of ES Module E:\workspace\_externals\mcsleepingserverstarter\node_modules\@jsprismarine\prismarine\node_modules\strip-json-comments\index.js from E:\workspace\_externals\mcsleepingserverstarter\node_modules\@jsprismarine\prismarine\dist\ban\BanManager.cjs.cjs not supported.
`

I also add to add brigadier to the package.json :
`    "@jsprismarine/prismarine": "^0.9.0",
    "@jsprismarine/bedrock-data": "^0.9.0",
    "@jsprismarine/brigadier": "^0.9.0",`

**Comment by enricoangelon** [comment]
> Hi, Again thanks for your quick answer, now it fails a bit further again : `E:\workspace_externals\mcsleepingserverstarter\node_modules@jsprismarine\prismarine\dist\ban\BanManager.cjs.cjs:6 const minifyJson = require('strip-json-comments'); ^
> 
> Error [ERR_REQUIRE_ESM]: require() of ES Module E:\workspace_externals\mcsleepingserverstarter\node_modules@jsprismarine\prismarine\node_modules\strip-json-comments\index.js from E:\workspace_externals\mcsleepingserverstarter\node_modules@jsprismarine\prismarine\dist\ban\BanManager.cjs.cjs not supported. `
> 
> I also add to add brigadier to the package.json : ` "@jsprismarine/prismarine": "^0.9.0", "@jsprismarine/bedrock-data": "^0.9.0", "@jsprismarine/brigadier": "^0.9.0",`

The lack of support for `.cjs` in one of our dependencies is causing the problem. We might consider adopting an alternative, maybe [strip-json-comments](https://www.npmjs.com/package/strip-json-comments). What do you think, @filiphsps?

**Comment by filiphsps** [comment]
> > Hi, Again thanks for your quick answer, now it fails a bit further again : `E:\workspace_externals\mcsleepingserverstarter\node_modules@jsprismarine\prismarine\dist\ban\BanManager.cjs.cjs:6 const minifyJson = require('strip-json-comments'); ^
> > Error [ERR_REQUIRE_ESM]: require() of ES Module E:\workspace_externals\mcsleepingserverstarter\node_modules@jsprismarine\prismarine\node_modules\strip-json-comments\index.js from E:\workspace_externals\mcsleepingserverstarter\node_modules@jsprismarine\prismarine\dist\ban\BanManager.cjs.cjs not supported. `
> > I also add to add brigadier to the package.json : ` "@jsprismarine/prismarine": "^0.9.0", "@jsprismarine/bedrock-data": "^0.9.0", "@jsprismarine/brigadier": "^0.9.0",`
> 
> The lack of support for `.cjs` in one of our dependencies is causing the problem. We might consider adopting an alternative, maybe [strip-json-comments](https://www.npmjs.com/package/@luxass/strip-json-comments). What do you think, @filiphsps?

Yep, or better yet switch to `json5` which natively supports comments (eg what `tsconfig.json` uses). Can potentially look into it.

**Comment by filiphsps** [comment]
We maybe also shouldn't require things like the ban manager to be included when a dependent only want access to the protocol and networking parts.

**Comment by filiphsps** [comment]
Alright, third time might be the charm. Please try version `0.10.0-unstable-20241027011822` @vincss and report back; I'll publish it properly together with a few dep bumps as soon as we've verified the `cjs` support.

(I would recommend moving to `esm` in the future as that's the industry standard and recommended happy path 🤠).

**Comment by vincss** [comment]
Hi thanks for this trial, we go another step further : 
``` javascript
F:\dev\mcsleepingserverstarter\node_modules\@jsprismarine\color-parser\dist\Colors.cjs.cjs:5
const chalk = require('chalk');
              ^

Error [ERR_REQUIRE_ESM]: require() of ES Module F:\dev\mcsleepingserverstarter\node_modules\@jsprismarine\color-parser\node_modules\chalk\source\index.js from F:\dev\mcsleepingserverstarter\node_modules\@jsprismarine\color-parser\dist\Colors.cjs.cjs not supported.
Instead change the require of index.js in F:\dev\mcsleepingserverstarter\node_modules\@jsprismarine\color-parser\dist\Colors.cjs.cjs to a dynamic import() which is available in all CommonJS modules.
    at Module.<anonymous> (F:\dev\mcsleepingserverstarter\node_modules\@jsprismarine\color-parser\dist\Colors.cjs.cjs:5:15) {
  code: 'ERR_REQUIRE_ESM'
}
```

I already try several times to switch to esm, but there is issue with the web-gui part, and the packaging to an single executable file. But it's still on my Todo list.

**Comment by filiphsps** [comment]
> Hi thanks for this trial, we go another step further :
> 
> ```js
> F:\dev\mcsleepingserverstarter\node_modules\@jsprismarine\color-parser\dist\Colors.cjs.cjs:5
> const chalk = require('chalk');
>               ^
> 
> Error [ERR_REQUIRE_ESM]: require() of ES Module F:\dev\mcsleepingserverstarter\node_modules\@jsprismarine\color-parser\node_modules\chalk\source\index.js from F:\dev\mcsleepingserverstarter\node_modules\@jsprismarine\color-parser\dist\Colors.cjs.cjs not supported.
> Instead change the require of index.js in F:\dev\mcsleepingserverstarter\node_modules\@jsprismarine\color-parser\dist\Colors.cjs.cjs to a dynamic import() which is available in all CommonJS modules.
>     at Module.<anonymous> (F:\dev\mcsleepingserverstarter\node_modules\@jsprismarine\color-parser\dist\Colors.cjs.cjs:5:15) {
>   code: 'ERR_REQUIRE_ESM'
> }
> ```
> 
> I already try several times to switch to esm, but there is issue with the web-gui part, and the packaging to an single executable file. But it's still on my Todo list.

Alright, try `0.10.0-unstable-20241027201936`.

**Comment by vincss** [comment]
It's alive ! :-D

The server is running and I can see it in the minecraft bedrock client : 

Sometimes, I got an error.
On client side  ( v1.21.41 ) : it says that the package is damaged.

And server side :
[22:31:55 WARN]: Packet 0x81 isn't implemented
[22:31:55 ERROR]: Packet violation 0x6, message: "BinaryStream read() incomplete
readNoHeader failed! packetId: 6", type: 0, severity: 2

Maybe it's because client side is v1.21.41 ?

In the readme you said: 

> JSPrismarine is made for Minecraft: Bedrock Edition, but there may potentially be plugins available to provide a Minecraft: Java Edition protocol implementation.
Is there such a plugin ?

Thanks :-)

**Comment by filiphsps** [comment]
> readNoHeader failed! packetId: 6", type: 0, severity: 2
We're currently using the latest bedrock-data and reporting the latest(-ish) protocol version while not actually having implemented it. I kind of had to bump it as a part of moving it to the monorepo. We still have a lot to figure out regarding handling protocol versioning optimally.

> In the readme you said:
Kind of, I experimented a bit way back on one of our older architectures as a way to validate and test part of the extensibility https://github.com/JSPrismarine/BigBrother . Right now the whole plugin subsystem has been ripped out along with a lot of other things to focus on improving and refactoring the fundamentals; of course neither me nor @enricoangelon have that much time to dedicate to jsp so that is probably just a pipe-dream.

**Comment by filiphsps** [comment]
I just merged the PR for publishing `v0.10.0`, which resolves this issue completely :^)

Let me know if there are any regressions in the future as we don't really test/validate `cjs` support 🤠
